lib/rubygems/path_support.rb

```diff
@@ -18,20 +18,35 @@ class Gem::PathSupport
   # Directory with spec cache
   attr_reader :spec_cache_dir # :nodoc:
 
+  ##
+  # Whether or `Gem.paths.home` defaulted to a user install.
+  attr_reader :auto_user_install
+
   ##
   #
   # Constructor. Takes a single argument which is to be treated like a
   # hashtable, or defaults to ENV, the system environment.
   #
   def initialize(env)
-    @home = env["GEM_HOME"] || Gem.default_dir
-
-    if File::ALT_SEPARATOR
-      @home = @home.gsub(File::ALT_SEPARATOR, File::SEPARATOR)
+    # Current implementation of @home, which is exposed as `Gem.paths.home`:
+    # 1. If `env["GEM_HOME"]` is defined in the environment: `env["GEM_HOME"]`.
+    # 2. If `Gem.default_dir` is writable OR it does not exist and it's parent
+    #    directory is writable: `Gem.default_dir`.
+    # 3. Otherwise: `Gem.user_dir`.
+
+    if env.key?("GEM_HOME")
+      @home = normalize_home_dir(env["GEM_HOME"])
+    elsif File.writable?(Gem.default_dir) || \
+          (!File.exist?(Gem.default_dir) && File.writable?(File.expand_path("..", Gem.default_dir)))
```

Comment:
I think if you extract a helper method `does_not_exist_but_parent_is_writable(<directory>)`, the code become self explainable and you don't really need the comment above?

Reply:
Also, `File.writable?` on directories seems to not do what one would expect:

```
$ mkdir foo
$ chmod -x foo
$ touch foo/bar
touch: cannot touch 'foo/bar': Permission denied
$ ruby -e 'puts File.writable?("foo")'
true
```

So, for directories, I think you need to check all permission bits: `File.writable? && File.readable? && File.executable?`.

Reply:
BTW, no idea how this works on Windows.

Reply:
Oh, right. idk about read permissions, but I do know that you need to be able to _write_ and _execute_ a directory to create it on Linux, and I believe most (if not all?) Unixes and Unix-likes.

Maybe this is why the original test for this was just `!File.exist?(Gem.default_dir)` instead of `!File.exist?(Gem.default_dir) && File.writable?(File.expand_path("..", Gem.default_dir))`.

I'll revisit this logic when I next work on this.

Reply:
Looks like this is the only remaining problem, of the things you brought up.

Reply:
Yes, there's also the refactoring I mentioned in the first comment in this review thread, but not a blocker at all!